BackEnd/source/server.js

```diff
@@ -0,0 +1,257 @@
+      customGameBoardValues[i][j] = null;
+    }
+  }
+  return customGameBoardValues;
+}
+
+function updateStateValues(newStateValues) {
+  state.gameBoardValues = newStateValues.gameBoardValues;
+  state.currentPlayer = newStateValues.currentPlayer;
+  state.gameBoardColumns = newStateValues.gameBoardColumns;
+  state.gameBoardRows = newStateValues.gameBoardRows;
+}
+
+function changePlayer(currentPlayerColour) { // PURE FUNCTION
+  if (currentPlayerColour === 'yellow') {
+    return 'red';
+  }
+  return 'yellow';
+}
+
+function placeCounter(columnIndex) {
+  for (let i = state.gameBoardValues[columnIndex].length - 1; i >= 0; i -= 1) {
+    if (state.gameBoardValues[columnIndex][i] == null) {
+      state.gameBoardValues[columnIndex][i] = state.currentPlayer;
+
+      state.currentPlayer = changePlayer(state.currentPlayer);
+
+      const counterPos = `${columnIndex}-${i}`;
+      return counterPos;
+    }
+  }
+  return null;
+}
+
+function checkColumnWin(columnIndex, gameBoardArray) { // PURE FUNCTION
+  let fourInARowCount = 0;
+  for (let i = gameBoardArray[columnIndex].length - 1; i >= 0; i -= 1) { // Loops through each row
+    if ((i + 1) >= 4 && gameBoardArray[columnIndex][i] !== null) {
+      fourInARowCount = 0;
+      for (let j = 1; j < 4; j += 1) {
+        const nextRowUpPiece = gameBoardArray[columnIndex][i - j];
+        if (gameBoardArray[columnIndex][i] !== nextRowUpPiece || nextRowUpPiece === null) {
+          break;
+        } else if (gameBoardArray[columnIndex][i] === nextRowUpPiece) {
+          fourInARowCount += 1;
+        }
+      }
+    } else {
+      break;
+    }
+    if (fourInARowCount >= 3) {
+      return gameBoardArray[columnIndex][i];
+    }
+  }
+  return null;
+}
+
+function checkRowWin(rowIndex, gameBoardArray) { // PURE FUNCTION
+  let fourInARowCount = 0;
+  for (let i = 0; i < gameBoardArray.length - 1; i += 1) {
+    if (gameBoardArray.length - i >= 4) {
+      fourInARowCount = 0;
+      for (let j = 1; j < 4; j += 1) {
+        const nextColumnPiece = gameBoardArray[i + j][rowIndex];
+        if (gameBoardArray[i][rowIndex] !== nextColumnPiece || nextColumnPiece === null) {
+          break;
+        } else if (gameBoardArray[i][rowIndex] === nextColumnPiece) {
+          fourInARowCount += 1;
+        }
+      }
+    } else {
+      break;
+    }
+    if (fourInARowCount >= 3) {
+      return gameBoardArray[i][rowIndex];
+    }
+  }
+  return null;
+}
+
+function checkRightDiagonalWin(colAndRowIndex, gameBoardArray) { // PURE FUNCTION
+  let fourInARowCount = 0;
+  const columnLength = gameBoardArray.length;
+
+  let columnPos = colAndRowIndex[0];
+  let rowPos = colAndRowIndex[1];
+  while (columnPos <= columnLength - 4 && rowPos >= 3) {
+    fourInARowCount = 0;
+    for (let i = 1; i < 4; i += 1) {
+      if (gameBoardArray[columnPos + i][rowPos - i] !== null) {
+        if (gameBoardArray[columnPos][rowPos] === gameBoardArray[columnPos + i][rowPos - i]) {
+          fourInARowCount += 1;
```

Comment:
This is the third time I've seen this logic now. And it's pretty complex. That means it probably needs to be broken out into a function somehow!